File: Cheapest-Flights-Within-K-Stops.java
1class Solution {
2    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) 
3    {
4             int[] dist = new int[n];
5             Arrays.fill(dist, Integer.MAX_VALUE);
6             dist[src]= 0;
7             for(int j=1;j<=k+1;j++)
8             {
9                int[] temp = Arrays.copyOf(dist, n);
10                for(int i=0;i<flights.length;i++)
11                {
12                    if(dist[flights[i][0]] != Integer.MAX_VALUE && dist[flights[i][0]] + flights[i][2] < temp[flights[i][1]])
13                        temp[flights[i][1]] = dist[flights[i][0]] + flights[i][2];
14                }
15                dist = temp;
16             }
17
18            return dist[dst] == Integer.MAX_VALUE ? -1 : dist[dst];
19
20
21    }
22}

File: Cheapest-Flights-Within-K-Stops.cpp
1class Solution {
2public:
3    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) 
4    {
5        vector<int> dist;
6        vector<vector<pair<int,int>>> adjList;
7        for(int i=0;i<n;i++)
8            dist.push_back(INT_MAX);
9        for(int i=0;i<n;i++)
10        {
11            vector<pair<int,int>> vec;
12            adjList.push_back(vec);
13        }
14        dist[src] = 0;
15        for(int i=0;i<flights.size();i++)
16        {
17            adjList[flights[i][0]].push_back({flights[i][1], flights[i][2]});
18        }  
19        for(int i=1;i<=k+1;i++)
20        {
21            vector<int> temp= dist;
22            for(int j=0;j<flights.size();j++)
23            {
24                int source = flights[j][0];
25                int dest = flights[j][1];
26                int weight = flights[j][2];
27                if(dist[source]!= INT_MAX && dist[source] + weight < temp[dest])
28                {
29                    temp[dest] = dist[source] + weight;
30                }            
31            }
32            dist= temp;
33        }
34        if(dist[dst] == INT_MAX)
35            return -1;
36        return dist[dst];
37    }
38};